### ISSUE TITLE
Pushing an image from a private registry

### ISSUE THREAD

**Issue by markns** [dokku/dokku#498]
I saw https://github.com/progrium/dokku/issues/313 which removed the ability to specify the $IMAGE in calls to build, release and deploy. 

In my scenario that's not helpful, since I am attempting the following workflow:
1. Run the build phase on a CI server
2. Push the image from CI to a private registry with a post-build plugin
3. Push the image to deployment environments by specifying the image name to the release and deploy phases.

Some organisations are quite averse to 'compiling in production', whereas the idea of pushing an image is quite appealing. Has anyone been thinking along similar lines?

Side question: is there a newsgroup, chatroom or otherwise that is more appropriate for these type of questions?

Thanks,
Mark

**Comment by josegonzalez** [comment]
Note: with the official [`tag` plugin](http://progrium.viewdocs.io/dokku/deployment/images/), you can now deploy images directly to dokku.

**Comment by ernoaapa** [comment]
@josegonzalez can you explain how it's possible deploy images directly to Dokku with `tag plugin`? Didn't find out this from the documentation.
Or should I use [dokku-registry](https://github.com/agco/dokku-registry) to do that? Thanks for help!

**Comment by josegonzalez** [comment]
The documentation is a bit spotty on how you would use it, but the example is there.

If you have an app called `playground`, it's dokku image name will be `dokku/playground`. You can tag images using docker directly, or you can tag the current release using the tag plugin:

```shell
 dokku tags:create playground MY_TAG
```

Now you have a tagged image that you can push/pull to wherever.

Once it is tagged, you can go back and deploy it as well:

```shell
dokku tags:deploy playground MY_TAG
```

Note that you can also pull tagged releases from other registries, and so long as the "repository" name is `dokku/APP_NAME`, you can deploy it easily, allowing you to build deploy pipelines with zero-downtime.

**Comment by chrisnicola** [comment]
@josegonzalez how to use use dokku tags to push and pull from other repositories. Reading through the docks it looks like you have to use `docker save` to pipe the image to the dokku server and then load it.

It would be awesome if you could configure dokku for access to your container registry (we use our own Gitlab instance) and then use simple dokku commands to deploy releases from it.

**Comment by josegonzalez** [comment]
@chrisnicola Please see our documentation on image workflows [here](http://dokku.viewdocs.io/dokku/deployment/methods/images/#image-workflows) - it was already linked to above.

**Comment by chrisnicola** [comment]
@josegonzalez it was, and I did read it and I'm trying now to integrate it into our CI process. I'm actually commenting that it is a bit of an awkward workflow involving piping the image to the dokku system's docker command via SSH. This means it doesn't work with dokku keys added via the ssh-keys command.

I feel the ability to release an image directly from a private repo would be very useful to be implemented directly in a dokku plugin.

I've noted here https://github.com/dokku/dokku/issues/2904 that the dokku registry plugin actually does support the kind of functionality I'm looking for. Sadly it appears disused and doesn't support 0.7+. I'm forking and patching it for my own usage, but I have proposed that this functionality be part of dokku itself, as I think the workflow of building images separately from deploying them makes a lot of sense.
